### DISCUSSION TITLE
Improving `l1l2` invariance measure

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#730] [Q&A]
A problem with `l1l2(x0, x1)` is measuring distance relative to `x0`, when there's no reason to prefer `x0` over `x1` as the reference:

  1. Equal time shifts (FDTS without FD)
  2. Equal frequency transpositions

`l1l2(x0, x1)` can differ considerably from `l1l2(x1, x0)` due to overall energy differences in `x0` & `x1`. In such cases, isn't below better? ([original discussion](https://github.com/kymatio/kymatio/discussions/711#discussioncomment-579329))

```
l1l2_mod(x0, x1) = _l1l2(x1 - x0) / ( (_l1l2(x0) + _l1l2(x1))/2 )
```

**Comment by lostanlen** [comment]
The only theorem we have is when `x1` is a deformed version of `x0`.
Section 2.5: https://www.di.ens.fr/~mallat/papiers/ScatCPAM.pdf

Typically, you might imagine that `x0` is in your training set while `x1` is in your test set.
In that case, it makes sense to make the denominator depend only on `x0` and not on `x1`.

**Reply by OverLordGoldDragon** [reply]
Yes, but the metric is flawed when we have no preference for reference but get notably different results with either. This 'adjustment' makes it robust to overall norm changes which helps when said changes are unrelated to what's being tested (e.g. freq shifts).

So this is more about improved statistics than matching the theorem.

**Reply by lostanlen** [reply]
> This 'adjustment' makes it robust to overall norm changes which helps when said changes are unrelated to what's being tested (e.g. freq shifts).

i don't expect frequency transposition to affect the energy of `x0`?

**Reply by OverLordGoldDragon** [reply]
Right. I also haven't reproduced this since, so no current need, but worth looking out for.

**Comment by OverLordGoldDragon** [comment]
Which will better discriminate echirps?

 1. `J, Q, J_fr, Q_fr = 10, 8, 4, 1`
 2. `J, Q, J_fr, Q_fr = 7, 8, 4, 1`

One can reason, 2 subsets 1, so 1 can't possibly do worse. Except... 2 is 7x better than 1:

```
J  | E_down/E_up
3  | 5.5
4  | 23.0
5  | 75.2
6  | 111.4
7  | 127.2
8  | 96.2
9  | 42.7
10 | 17.7
```
that is, according to the L2 metric. The explanation is, greater `J` introduces more lower-frequency `psi_t`, which fare poorer at detecting narrow FDTS (likes of echirp), so they increase the denominator (norm of reference) but not numerator (norm of difference).

Yet, from a classifier's perspective, if feeding joint slices separately, discriminative power isn't lost. Hence another reason the metric is flawed.

A more accurate measure is obtained by comparing energies per-coefficient, as via `energy_profile_jtfs`:

![image](https://user-images.githubusercontent.com/16495490/120322894-fb127900-c2f5-11eb-9efc-62358c476f51.png)

and greater `J` would simply correspond to more zeros in the plot. The remaining question is how to aggregate into a single score.

<details>
  <summary><b>code</b></summary>

```python
import numpy as np
from kymatio import TimeFrequencyScattering1D

def _l2(x):
    return np.sqrt(np.sum(np.abs(x)**2))

def l2(x0, x1):
    """Coeff distance measure; Eq 2.24 in
    https://www.di.ens.fr/~mallat/papiers/ScatCPAM.pdf
    """
    return _l2(x1 - x0) / _l2(x0)

def energy(x):
    return np.sum(np.abs(x)**2)


def echirp(N, fmin=.1, fmax=None, tmin=0, tmax=1):
    """https://overlordgolddragon.github.io/test-signals/ (bottom)"""
    fmax = fmax or N // 2
    t = np.linspace(tmin, tmax, N)

    a = (fmin**tmax / fmax**tmin) ** (1/(tmax - tmin))
    b = fmax**(1/tmax) * (1/a)**(1/tmax)

    phi = 2*np.pi * (a/np.log(b)) * (b**t - b**tmin)
    return np.cos(phi)

#%%###########################################################################
print("J  | E_down/E_up")
for J in range(3, 11):
    N = 2048
    x = echirp(N)
    
    jtfs = TimeFrequencyScattering1D(shape=N, J=J, Q=8, J_fr=4, Q_fr=1,
                                     average_fr=True, out_3D=True,
                                     out_type='dict:array', frontend='numpy')
    Scx = jtfs(x)
    
    E_up   = energy(Scx['psi_t * psi_f_up'])
    E_down = energy(Scx['psi_t * psi_f_down'])
    print("{:<2} | {:.1f}".format(J, E_down / E_up))
```
</details>

**Reply by lostanlen** [reply]
Point taken.
Are you doing global average pooling in time before doing the comparison or not? if not, it's not that surprising that too much averaging in time (large J) reduces discriminability. am i missing something?

**Reply by OverLordGoldDragon** [reply]
No global averaging. Averaging doesn't affect FDTS discriminability, no? It simply linearly combines existing correlations, and the metric aggregates over entire joint slices.

Yet, I compared with and without time or frequency averaging, with maximum oversampling; neither has clear edge over the other, but time averaging is much more pronounced for large `J` (showing time averaged only):

```
     averaged    | unaveraged
J  | E_down/E_up |
3  | 4.3         | 4.1
4  | 23.4        | 22.1
5  | 72.8        | 65.2
6  | 110.5       | 97.4
7  | 121.7       | 112.5
8  | 94.4        | 105.0
9  | 42.7        | 76.1
10 | 19.1        | 57.3
```
Spinned coeffs, semi-flattened, `J=10`:

![image](https://user-images.githubusercontent.com/16495490/120847664-da138780-c584-11eb-8054-00d51995eebf.png)

Both up and down are more pronounced in averaged, but seems the increase in denominator wins.

**Reply by OverLordGoldDragon** [reply]
So lowpass is indeed to blame, but in a strange way: interaction with `n2`.

The last `n2` slices, originally omitted from visuals per faulty code, make it clear: lowpass correlates strongly with high `n2` (high `j2`). Last unaveraged's up spin is hardly visible, but its averaging is about half of down's averaging. Averaged ups diminish toward lower `n2`.

In contrast, the strong FDTS correlations are relatively narrow in time, hence reduced more by lowpass than the weak correlations that are spread out in time. Hence the growth in denominator (ups) indeed wins. With last two `n2` omitted, the energy ratios of unaveraged and averaged are 104.1 and 62.9, much closer than before.

It seems fair enough to call this a reduction in discriminability - but again the plain ratios hid the dependence on `n2` that'd be revealed with a per-coefficient comparison. Per-coefficient is also robust to outliers strongly swaying the norm of the reference (denominator). (I currently lack code to run these comparisons, am yet to bring existing funcs up to date)

More definitively:

```
J  | T=2**3 | T=2**10
3  | 4.3    | 3.1
4  | 23.1   | 15.8
5  | 67.8   | 42.6
6  | 100.2  | 63.3
7  | 114.9  | 71.8
8  | 106.5  | 63.6
9  | 76.6   | 34.8
10 | 57.5   | 19.1
```

**Reply by OverLordGoldDragon** [reply]
Much clearer:

![image](https://user-images.githubusercontent.com/16495490/121051625-a4f87680-c7ca-11eb-93d6-99e404c4a799.png)

Note: coeff ordering reversed now, low to high freq

**Comment by OverLordGoldDragon** [comment]
## `width` computation notes

Briefly; notes on method: `kymatio.scattering1d.filter_bank.compute_temporal_width()`.

### Complete decay

`width == T`

<img src="https://user-images.githubusercontent.com/16495490/126065755-ee93f8e4-005d-4b78-8bbf-47f73ccc3c5a.png" width="450">


### Incomplete decay

Results in `width > T`:

![image](https://user-images.githubusercontent.com/16495490/126065639-3f5bd7a8-16c3-43fa-b2ff-a54b808659f1.png)

> for complete decay it's clear, it's the same L2 distance when shift == T on impulse response (or randn)
incomplete decay = N < 16*T
or to be precise, phi_t.min() / phi_t.max() > criterion_amplitude = 1e-3
this is what my bunch of blue curves show
for complete decay, shift == T is attained at same L2 - for incomplete decay, the distance is lesser, i.e. invariance is greater, the relationship breaks down

> The criteria so far are, that width == T for complete decay, and for global averaging, but for everything between, width > T

### L2 vs T, fixed N | no unpad

> L2 reldist vs shift for l2(phi_t, roll(phi_t, shift)), where each blue line is a different T, and to right of purple line are phis sampled at insufficient length

> red lines are shift == T, and they all meet at same L2 up until purple line, then decay, indicating that T imposes greater invariance than it advertises, as is expected from incomplete decay

<img src="https://user-images.githubusercontent.com/16495490/126065694-4fbc2099-8b83-4c4f-8393-30a13d381934.png" width="500">

### L2 vs N, fixed T | no unpad

<img src="https://user-images.githubusercontent.com/16495490/126065859-e24cc8e8-a0af-446d-959e-09bd1c77057b.png" width="450">

N vs L2 for same T, at maximum shift

<img src="https://user-images.githubusercontent.com/16495490/126065875-6e475d62-726e-4073-b2a2-d5e1d0955706.png" width="450">

### L2 failure

Euclidean distance claims left is "closer" to bottom than right is, in terms of distance between blue and orange:

![image](https://user-images.githubusercontent.com/16495490/126065556-4d952bd2-5e1e-4d2d-b0b9-629ef70cfca1.png)

This might be due to the blue/orange norms being different, where scattering theorems assume equal norms, but that renders the measure inapplicable for the case of incomplete decay, where we necessarily trim in unpadded case.

### Algorithm result, L2, integer T

Idealized for `T` sampled at integers; unstable for non-integers

<img src="https://user-images.githubusercontent.com/16495490/126065817-2cc2241d-d29d-420d-a088-ba8123bec162.png" width="450">

### Algorithm result, inner product, sweep T

<img src="https://user-images.githubusercontent.com/16495490/126065991-2f7ee99e-0ad1-4db2-a2a0-4f4ea7d903e1.png" width="450">

Very close to L2 measure for convenient "sweep T", but requires correction steps for general T.

The `fast=True` algorithm:

<img src="https://user-images.githubusercontent.com/16495490/126066003-d3f43e85-0593-4286-b506-163f5e4ad7c4.png" width="450">

The quadratic interpolation isn't quite quadratic since it takes as input a quantity nonlinearly related to `T`:

<img src="https://user-images.githubusercontent.com/16495490/126066186-22b47c2f-2ece-4d37-b55c-b7a28771f405.png" width="450">

It'll be made exactly quadratic if `ratio` is linearized with respect to `T`, but I've not found the inverting operator.

**Reply by lostanlen** [reply]
what are you trying to do here?

**Reply by lostanlen** [reply]
the workflow i was imagining was:

1) define invariance in some abstract way that is independent of the specific shape of phi  (maybe L2 distance = 0.5 or something)
2) define "complete decay" in some abstract way that is independent of specific shape of phi (maybe your 1e-3 inequality)
3) define phi = Gaussian. (right now this is our only supported shape for local averaging)
4) let the user decide T
5) define width depending on T and phi shape to ensure your definition of invariance (see 1)
6) define N depending on width and phi shape to ensure your definition of complete decay (see 2)
7) design phi of determined shape (see 3), width (see 5), and N (see 6)

are you following this workflow or a different one? what is your definition of T and width?

**Reply by lostanlen** [reply]
papers on scattering are based on the notion that the value of T is a quantitative measure of invariance so i would assume that your definition of invariance supports T as a continuous parameter

then it's a matter of designing the phi from the knowledge of T to fit this definition

**Reply by OverLordGoldDragon** [reply]
https://github.com/lostanlen/kymatio/blob/fix-672/kymatio/scattering1d/filter_bank.py#L1628

**Reply by lostanlen** [reply]
Where does this value of `sigma0=0.1` come from and why do you take it for granted? (as default value)

The determination seems backward to me. You start from a filter `p_f` and measure its "width". Wouldn't it be more direct to deduce the filter from a user-specified width?

What is your workflow?

**Reply by OverLordGoldDragon** [reply]
`sigma0=.1` is Kymatio's hard-coded default value. The user doesn't get to specify `width`, only `J`, and the two are related via `sigma0`, as `J` controls `sigmas`, which are set via `sigma0`.

`sigma0` can be exposed to the user, but that would not change anything about `width` computation:

> The interpretation is, "width" of `p_f` is the width of the (properly decayed)
 Gaussian lowpass that imposes the same amount of invariance that `p_f` does.

> Parameters defining the Gaussian lowpass used as reference for computing `width`.
That is, `width` is defined *in terms of* this Gaussian.

**Reply by OverLordGoldDragon** [reply]
The idea is, whatever user specifies as `T` *is* the `width`, _as long as the filter fully decays_. The matter then is to determine `width` from `T` when there _isn't_ complete decay, by defining `width` in terms of filters that _do_ fully decay (which will have more samples).

**Reply by lostanlen** [reply]
so in the practical setting (incomplete decay), does the user-specified value for invariance (`T`) end up being the actual amount of invariance or not? seems like not from your last message, but i don't see why that should be the case. People care about defining the actual invariance, not the invariance "as long as the filter fully decays"

**Reply by OverLordGoldDragon** [reply]
Yes, I was thinking maybe `T` should specify the actual width - but there'd be no clean way to do the same to the wavelets, where `J` would specify the "actual" scale: it'd create dependence on `sigma_max`, which already depends on `r_psi` and `Q`. Several other things go wrong with insufficient decay (see [here](https://github.com/kymatio/kymatio/pull/718#issuecomment-882061860)), and these do not have correction steps.

`T == width` should hold in most practical settings, however; padding by one factor less than maximum still keeps it pretty close (see plots), but as long as `J` or `log2_T` aren't `== log2_N`, we can afford to pad by x4 if `== log2_N - 1` or x2 if `== log2_N - 2`.

**Reply by OverLordGoldDragon** [reply]
Basically, the expectation to set is - "if you can't pad sufficiently, expect distortions." If a user favors speed over reliability, that's a tradeoff they're free to make.

**Reply by lostanlen** [reply]
> Basically, the expectation to set is - "if you can't pad sufficiently, expect distortions."

this is a very reasonable expectation indeed. in fact, the same happens in Conv1D.

>  If a user favors speed over reliability, that's a tradeoff they're free to make.

so, can you recap the workflow? and explain where the tradeoff happens?

how many degrees of freedom do you want to allow on the part of the user?

**Reply by OverLordGoldDragon** [reply]
### Reasoning 'summary'

I seek to quantify "width" of a filter such that it directly reflects the amount of time-shift invariance it imposes via convolution. One measure of "invariance" is relative L2 (Euclidean) distance.

The algorithm is described in the docstring, but to recap: (1) compute L2 of input filter with itself at maximum shift; (2) compute L2 of several filters between `T_min` and `T_max` (chosen to guarantee containing actual `T`); (3) `width` is the `T` that most closely matches the result of (1). As long as the filter fully decays, this approach yields `T == width` for every `T`.

Problems ensue with incomplete decay. Unpadding changes the compared norms and breaks a monotonic relationship between `T` and invariance. The blue curves above are from circular correlation: with unpadding, there's an intermediate peak before leveling off into an asymptote, which makes the relation between `T` and `L2` not one-to-one, thus we cannot determine `T` from `L2` (rather `width`; define `T` as the continuous-time parameter used to generate the filter). See "L2 failure" above for an example.

Like a quadratic with "invalid" solutions, there exist correction steps to rule out the incorrect match, but it's both computationally expensive and complicated _even if_ assuming Gaussian-only. A fully general filter shape (which I attempted) is simply out of question at this rate. (I don't claim there's no nice solution, but I simply haven't found it)

<hr>

> and explain where the tradeoff happens?

A [bunch of places](https://github.com/kymatio/kymatio/pull/718#issuecomment-882061860)... Here specifically, the tradeoff is an uncertainty on amount of imposed invariance (unless my measure makes sense to the user). The other tradeoff is inaccuracy with `fast=True`, which approximates `fast=False` (see above), but the difference is small. Again, I'm open to the idea of having `T` specify the actual `width`, but I can't see this happening for "scale" of wavelets without massive complications.

> how many degrees of freedom do you want to allow on the part of the user?

There's `T`, `J`, and `max_pad_factor`. In most practical settings, we'll have both `log2_T` and `J` that are `< log2_N - 1`, and there is no problem. If this doesn't hold, the user can specify `max_pad_factor=None` and `T` will exactly specify `width`. _However_ - run `test_global_averaging()` with `T = N-1 -> T = N-50`, and you'll see that the relative L2 distance between it and `T == N` is on the order of `1e-3`, which confirms the "saturation" suggested by my measure.

> will the user ever call compute_temporal_width? what purpose does it have?

No, just like they'll never call `compute_temporal_support()`.

> also, the value sigma0=0.1 is clearly ad hoc and should be removed from the definitions.

Hopefully above clarifies the dependence on `sigma0`. If the user is unhappy with `0.1`, we can expose it and the value will propagate to `compute_temporal_width()`. I'd like nothing more than a reference-independent measure, but this works well enough and further effort isn't worthwhile (unless you have a suggestion).
